akka-persistence-testkit/src/main/scala/akka/persistence/testkit/scaladsl/PersistenceTestKit.scala

```diff
@@ -0,0 +1,507 @@
+    extends CommonTestKitOps[(SnapshotMetadata, Any), SnapshotOperation]
+    with PolicyOpsTestKit[SnapshotOperation]
+    with ExpectOps[(SnapshotMetadata, Any)]
+    with HasStorage[SnapshotOperation, (SnapshotMetadata, Any)] {
+  require(
+    Try(Persistence(system).journalFor(PersistenceTestKitSnapshotPlugin.PluginId)).isSuccess,
+    "The test persistence plugin for snapshots is not configured.")
+
+  import SnapshotTestKit._
+
+  override protected val storage: SnapshotStorage = SnapshotStorageEmulatorExtension(system)
+
+  private val settings = Settings(system)
+
+  override private[testkit] val probe = TestProbe()
+
+  override private[testkit] val pollInterval: FiniteDuration = settings.pollInterval
+
+  override private[testkit] val maxTimeout: FiniteDuration = settings.assertTimeout
+
+  override private[testkit] val Policies = SnapshotStorage.SnapshotPolicies
+
+  override def failNextNPersisted(persistenceId: String, n: Int, cause: Throwable): Unit =
+    failNextNOpsCond((pid, op) => pid == persistenceId && op.isInstanceOf[WriteSnapshot], n, cause)
+
+  override def failNextNPersisted(n: Int, cause: Throwable): Unit =
+    failNextNOpsCond((_, op) => op.isInstanceOf[WriteSnapshot], n, cause)
+
+  override def failNextNReads(n: Int, cause: Throwable): Unit =
+    failNextNOpsCond((_, op) => op.isInstanceOf[ReadSnapshot], n, cause)
+
+  override def failNextNReads(persistenceId: String, n: Int, cause: Throwable): Unit =
+    failNextNOpsCond((pid, op) => pid == persistenceId && op.isInstanceOf[ReadSnapshot], n, cause)
+
+  override def failNextNDeletes(n: Int, cause: Throwable): Unit =
+    failNextNOpsCond((_, op) => op.isInstanceOf[DeleteSnapshot], n, cause)
+
+  override def failNextNDeletes(persistenceId: String, n: Int, cause: Throwable): Unit =
+    failNextNOpsCond((pid, op) => pid == persistenceId && op.isInstanceOf[DeleteSnapshot], n, cause)
+
+  /**
+   * Persist `elems` pairs of (snapshot metadata, snapshot payload) into storage.
+   */
+  def persistForRecovery(persistenceId: String, elems: immutable.Seq[(SnapshotMeta, Any)]): Unit =
+    elems.foreach {
+      case (m, p) =>
+        storage.add(persistenceId, (SnapshotMetadata(persistenceId, m.sequenceNr, m.timestamp), p))
+        addToIndex(persistenceId, 1)
+    }
+
+  /**
+   * Persist a pair of (snapshot metadata, snapshot payload) into storage.
+   */
+  def persistForRecovery(persistenceId: String, elem: (SnapshotMeta, Any)): Unit =
+    persistForRecovery(persistenceId, immutable.Seq(elem))
+
+  /**
+   * Retrieve snapshots and their metadata from storage by persistence id.
+   */
+  def persistedInStorage(persistenceId: String): immutable.Seq[(SnapshotMeta, Any)] =
+    storage
+      .read(persistenceId)
+      .map(_.map(m => (SnapshotMeta(m._1.sequenceNr, m._1.timestamp), m._2)))
+      .getOrElse(Vector.empty)
+
+  override private[testkit] def reprToAny(repr: (SnapshotMetadata, Any)) = repr._2
+
+}
+
+object SnapshotTestKit {
```

Comment:
we don't guarantee binary compatibility for testkits so we have some freedom to change, but could be good to highlight that this is a new module that is more likely to change by adding the `@ApiMayChange` annotation at a few strategic places